src/content/en/fundamentals/engage-and-retain/push-notifications/web-push-protocol.md

```diff
@@ -0,0 +1,613 @@
+
+    Authorization: 'WebPush <JWT Info>.<JWT Payload>.<Signature>'
+    
+The Web Push Protocol also states the public application server key must be
+sent in the `Crypto-Key` header as a URL safe base64 encoded string with
+`p256ecdsa=` prepended to it.
+
+    Crypto-Key: p256ecdsa=<URL Safe Base64 Public Application Server Key>
+    
+## The Payload Encryption
+
+Next let's look at how we can send a payload with a push message so that our when our web app
+receives a push message, it can access the data it receives.
+
+A common question that arises from any who's used other push services is why does the web push
+payload need to be encrypted? With native apps, push messages can send data as plain text.
+
+Part of the beauty of web push is that because all push services use the same API (the web push
+protocol), developers don't have to care who the push service is. We can make a request in the
+right format and expect a push message to be sent. The downside of this is that developers
+could conceivably send messages to a push service that isn't trustworthy. By encrypting the
+payload, push service can't read the data thats sent, only the browser can decrypt the
+information. This protects the users data.
+
+The encryption of the payload is defined in the [Message Encryption
+spec](https://tools.ietf.org/html/draft-ietf-webpush-encryption).
+
+Before we look at the specific steps to encrypt our payload for a PushSubscription, we should
+cover some techniques that'll be used during the encryption process (Massive H/T to Mat Scales
+for his excellent article on push encryption).
+
+### ECDH and HKDF
+
+Both ECDH and HKDF are used throughout the encryption process and offer benefits for the
+purpose of encrypting information.
+
+#### ECDH: Elliptic Curve Diffie-Hellman Key Exchange
+
+Imagine you have two people who want to share information, Alice and Bob. Both
+Alice and Bob have their own public and private keys. With ECDH, Alice and Bob
+share their public keys with each other. The useful property of ECDH is that
+Alice can use her private key and Bob's public key to create secret value 'X'.
+Bob can then do the same, taking his private key and Alice's public key to
+create the same secret value 'X'.
+
+Now Bob and Alice can use 'X' to encrypt and decrypt messages between them.
+
+ECDH, to the best of my knowledge, defines the properties of curves which allow this "feature"
+of making a shared secret 'X'.
+
+This is a high level explanation of ECDH, if you want to learn more [I recommend checking out
+this video](https://www.youtube.com/watch?v=F3zzNa42-tQ).
+
+In terms of code; most languages / platforms come with libraries to make it
+easy to generate these keys.
+
+In node we'd do the following:
+
+    const keyCurve = crypto.createECDH('prime256v1');
+    keyCurve.generateKeys();
+
+    const publicKey = keyCurve.getPublicKey();
+    const privateKey = keyCurve.getPrivateKey();
+    
+#### HKDF: HMAC Based Key Deriviation Function
```

Comment:
I'm sure I've commented before that everything other than level 1 should be sentence case. So: 

#### HKDF: HMAC based key deriviation function

Please do this throughout the project.